Implement GameWindowManager and W3DGameWindowManager

jonwil opened the pull request.

jonwil committed:
Implement GameWindowManager and W3DGameWindowManager

xezon commented on src/platform/w3dengine/client/gui/gadget/w3dcheckbox.cpp:
```diff
@@ -13,3 +13,17 @@
+void W3D_Gadget_Check_Box_Draw(GameWindow *check_box, WinInstanceData *data)
+{
+#ifdef GAME_DLL
+    Call_Function<void, GameWindow *, WinInstanceData *>(0x007CDDF0, check_box, data);
```
No PICK_ADDRESS() here? Several times in this review.

jonwil replied:
The W3D_Gadget_xxx functions don't exist in WB, only in game exe.

xezon commented on src/hooker/hooker.h:
```diff
@@ -131,6 +131,11 @@ inline void Hook_Func(uintptr_t in, uintptr_t out)
 
+template<typename T> void Hook_Memory(uintptr_t in, T out)
+{
+    WriteProcessMemory(GetCurrentProcess(), (LPVOID)in, &out, 4, nullptr);
```
`sizeof(T)`

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.h:
```diff
@@ -36,6 +36,33 @@ struct _TextData;
+    MSG_BOX_YES = 1,
+    MSG_BOX_NO = 2,
+    MSG_BOX_OK = 8,
+    MSG_BOX_CANCEL = 4,
```
Can fix order here.

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.h:
```diff
@@ -36,6 +36,33 @@ struct _TextData;
+    WIN_ERR_INVALID_PARAMETER = -3,
+    WIN_ERR_MOUSE_CAPTURED = -4,
+    WIN_ERR_KEYBOARD_CAPTURED = -5,
+    WIN_ERR_OUT_OF_WINDOWS = -6,
```
Can fix order here.

jonwil replied:
Would prefer to keep things in this order (i.e. from 0 down to -6) because of how things line up in the code.

xezon commented on src/game/client/gui/gamewindowglobal.cpp:
```diff
@@ -12,4 +12,99 @@
+
+void GameWindowManager::Win_Fill_Rect(int color, float width, int start_x, int start_y, int end_x, int end_y)
+{
+    g_theDisplay->Draw_Fill_Rect(start_x, start_y, end_x - start_x, end_y - start_y, color);
```
`width` is unused?

jonwil replied:
Yes its unused.

xezon commented on src/game/client/gui/gamewindowglobal.cpp:
```diff
@@ -12,4 +12,99 @@
 #include "gamewindowmanager.h"
+#include "image.h"
+
+void GameWindowManager::Win_Draw_Image(const Image *image, int start_x, int start_y, int end_x, int end_y, int color)
```
Perhaps call all these `left`, `top`, `right`, `bottom` instead of start x/y end x/y.

In wnd files this is is their naming convention.

xezon commented on src/game/client/gui/gamewindowglobal.cpp:
```diff
@@ -12,4 +12,99 @@
+    g_theDisplay->Draw_Fill_Rect(start_x, start_y, end_x - start_x, end_y - start_y, color);
+}
+
+void GameWindowManager::Win_Open_Rect(int color, float width, int start_x, int start_y, int end_x, int end_y)
```
I suspect `width` refers to border width. Perhaps name it accordingly to avoid confusion.

xezon commented on src/game/client/gui/gamewindowglobal.cpp:
```diff
@@ -12,4 +12,99 @@
+
+    if (height != nullptr) {
+        *height = 0;
+    }
```
Isn't there something missing in this function? Looks odd.

jonwil replied:
This is what original looks like.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -24,11 +26,43 @@
+        delete g_theTransitionHandler;
+    }
+
+    g_theTransitionHandler = nullptr;
```
Can be in scope above or removed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+}
+
+WindowMsgHandledType Pass_Selected_Buttons_To_Parent_System(
+    class GameWindow *window, unsigned int msg, unsigned int mData1, unsigned int mData2)
```
data1, data2

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+}
+
+WindowMsgHandledType Pass_Messages_To_Parent_System(
+    class GameWindow *window, unsigned int msg, unsigned int mData1, unsigned int mData2)
```
data1, data2

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+
+    GameWindow *parent = window->Win_Get_Parent();
+
+    if (parent) {
```
!= nullptr

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+    if (msg == GBM_SELECTED || msg == GBM_SELECTED_RIGHT || msg == GBM_MOUSE_ENTERING || msg == GBM_MOUSE_LEAVING
+        || msg == GEM_EDIT_DONE) {
+        GameWindow *parent = window->Win_Get_Parent();
+        if (parent) {
```
!= nullptr

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+void GameWindowManager::Add_Window_To_Parent(GameWindow *window, GameWindow *parent)
+{
+    if (parent != nullptr) {
+        window->m_prev = 0;
```
nullptr

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+{
+    if (parent != nullptr) {
+        window->m_prev = 0;
+        window->m_next = 0;
```
nullptr

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+    }
+}
+
+void GameWindowManager::Hide_Windows_In_Range(GameWindow *base_window, int first, int last, bool state)
```
bool hide

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+    }
+}
+
+void GameWindowManager::Enable_Windows_In_Range(GameWindow *base_window, int first, int last, bool state)
```
bool enable

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+
+int GameWindowManager::Win_Set_Focus(GameWindow *window)
+{
+    bool b2;
```
No default value here? Looks as b2 is not used other than passing into function below.

jonwil replied:
No default value. Code that handles GWM_INPUT_FOCUS will fill b2 in with some value (which means b2 needs to be passed in the way it is) but the code in Win_Set_Focus doesn't then care about the value that gets put into b2.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+{
+    WinInputReturnCode ret = WIN_INPUT_NOT_USED;
+
+    if (m_keyboardFocus && key) {
```
!= nullptr, != 0

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+    IRegion2D new_region;
+    ICoord2D grab_size;
+    unsigned int packed_mouse_coords =
+        (static_cast<unsigned short>(mouse_pos->y) << 16) | static_cast<unsigned short>(mouse_pos->x);
```
ushort << 16 will not be 0 ?

jonwil replied:
This is producing the correct values so it seems to be OK.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+    WindowMsgBoxCallbackFunc no_callback,
+    WindowMsgBoxCallbackFunc ok_callback,
+    WindowMsgBoxCallbackFunc cancel_callback,
+    bool use_other)
```
Perhaps `bool quit`

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+        }
+
+        GameWindow *parent = nullptr;
+        Utf8String str2 = str;
```
window_name

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+        Win_Set_Focus(parent);
+
+        if (width > 0 && height > 0) {
+            int i1;
```
window_width

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+
+        if (width > 0 && height > 0) {
+            int i1;
+            int i2;
```
window_height

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+            int i1;
+            int i2;
+            parent->Win_Get_Size(&i1, &i2);
+            float f3 = (float)width / (float)i1;
```
x_rescale_ratio

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+            int i2;
+            parent->Win_Get_Size(&i1, &i2);
+            float f3 = (float)width / (float)i1;
+            float f4 = (float)height / (float)i2;
```
y_rescale_ratio

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+            return nullptr;
+        }
+
+        lb_data->m_columnWidth = new int[lb_data->m_columns];
```
Is this `m_columnWidth` ever deleted anywhere?

jonwil replied:
Probably in code we haven't yet cloned.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+        if (tab_control != nullptr) {
+            _TabControlData *tc_data = new _TabControlData;
+            memcpy(tc_data, data, sizeof(_TabControlData));
+            tab_control->Win_Set_User_Data(tc_data);
```
Is the window user data ever deleted anywhere?

jonwil replied:
Probably in code we haven't yet cloned.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+            status &= ~(WIN_STATUS_BORDER | WIN_STATUS_HIDDEN);
+            int font_height = g_theWindowManager->Win_Font_Height(combo_box->Win_Get_Font());
+            int i1;
+            int i2;
```
i1, i2 are never used other than the value assignment below. Mistake?

jonwil replied:
No, this is what the code looks like. I could remove the not-needed code but I would rather keep it matching the original for now.

xezon replied:
Perhaps add a comment for it to clarify why it is there, but unused.

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+
+            progress_bar->Win_Set_Draw_Func(draw);
+            progress_bar->Win_Set_Owner(parent);
+            this->Assign_Default_Gadget_Look(progress_bar, font, assign_visuals);
```
`this->` can be omitted?

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+    data->m_conCharPos = 0;
+    data->m_receivedUnichar = 0;
+
+    if (data->m_maxTextLen >= 256) {
```
`> 256`

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                Gadget_Slider_Set_Thumb_Enabled_Image(gadget, Win_Find_Image("HSliderThumbEnabled"));
+                Gadget_Slider_Set_Thumb_Enabled_Color(gadget, Gadget_Slider_Get_Enabled_Color(gadget));
+                Gadget_Slider_Set_Thumb_Enabled_Border_Color(gadget, Gadget_Slider_Get_Enabled_Border_Color(gadget));
+                Gadget_Slider_Set_Thumb_Enabled_Selected_Image(gadget, Win_Find_Image("HSliderThumbEnabled"));
```
Not "HSliderThumbEnabledSelected" ?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                Gadget_Slider_Set_Thumb_Disabled_Image(gadget, Win_Find_Image("HSliderThumbDisabled"));
+                Gadget_Slider_Set_Thumb_Disabled_Color(gadget, Gadget_Slider_Get_Disabled_Color(gadget));
+                Gadget_Slider_Set_Thumb_Disabled_Border_Color(gadget, Gadget_Slider_Get_Disabled_Border_Color(gadget));
+                Gadget_Slider_Set_Thumb_Disabled_Selected_Image(gadget, Win_Find_Image("HSliderThumbDisabled"));
```
Not "HSliderThumbDisabledSelected" ?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                Gadget_Slider_Set_Thumb_Enabled_Image(gadget, Win_Find_Image("VSliderThumbEnabled"));
+                Gadget_Slider_Set_Thumb_Enabled_Color(gadget, Gadget_Slider_Get_Enabled_Color(gadget));
+                Gadget_Slider_Set_Thumb_Enabled_Border_Color(gadget, Gadget_Slider_Get_Enabled_Border_Color(gadget));
+                Gadget_Slider_Set_Thumb_Enabled_Selected_Image(gadget, Win_Find_Image("VSliderThumbEnabled"));
```
Not "VSliderThumbEnabledSelected"?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                Gadget_Slider_Set_Thumb_Disabled_Image(gadget, Win_Find_Image("VSliderThumbDisabled"));
+                Gadget_Slider_Set_Thumb_Disabled_Color(gadget, Gadget_Slider_Get_Disabled_Color(gadget));
+                Gadget_Slider_Set_Thumb_Disabled_Border_Color(gadget, Gadget_Slider_Get_Disabled_Border_Color(gadget));
+                Gadget_Slider_Set_Thumb_Disabled_Selected_Image(gadget, Win_Find_Image("VSliderThumbDisabled"));
```
Not "VSliderThumbDisabledSelected"?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                    Gadget_Slider_Set_Enabled_Small_Repeating_Centre_Image(
+                        lb_slider, Win_Find_Image("VSliderLargeEnabledSmallRepeatingCenter"));
+                    Gadget_Slider_Set_Thumb_Enabled_Image(lb_slider, Win_Find_Image("VSliderLargeThumbEnabled"));
+                    Gadget_Slider_Set_Thumb_Enabled_Selected_Image(lb_slider, Win_Find_Image("VSliderLargeThumbEnabled"));
```
Not "VSliderLargeThumbEnabledSelected"?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                    Gadget_Slider_Set_Disabled_Small_Repeating_Centre_Image(
+                        lb_slider, Win_Find_Image("VSliderLargeDisabledSmallRepeatingCenter"));
+                    Gadget_Slider_Set_Thumb_Disabled_Image(lb_slider, Win_Find_Image("VSliderLargeThumbDisabled"));
+                    Gadget_Slider_Set_Thumb_Disabled_Selected_Image(lb_slider, Win_Find_Image("VSliderLargeThumbDisabled"));
```
Not "VSliderLargeThumbDisabledSelected"?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                    Gadget_Slider_Set_Hilite_Small_Repeating_Centre_Image(
+                        lb_slider, Win_Find_Image("VSliderLargeHiliteSmallRepeatingCenter"));
+                    Gadget_Slider_Set_Thumb_Hilite_Image(lb_slider, Win_Find_Image("VSliderLargeThumbHilite"));
+                    Gadget_Slider_Set_Thumb_Hilite_Selected_Image(lb_slider, Win_Find_Image("VSliderLargeThumbHilite"));
```
Not "VSliderLargeThumbHiliteSelected" ?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                        Gadget_Slider_Set_Thumb_Enabled_Image(
+                            cb_list_box_slider, Win_Find_Image("VSliderLargeThumbEnabled"));
+                        Gadget_Slider_Set_Thumb_Enabled_Selected_Image(
+                            cb_list_box_slider, Win_Find_Image("VSliderLargeThumbEnabled"));
```
Not "VSliderLargeThumbEnabledSelected" ?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                        Gadget_Slider_Set_Thumb_Disabled_Image(
+                            cb_list_box_slider, Win_Find_Image("VSliderLargeThumbDisabled"));
+                        Gadget_Slider_Set_Thumb_Disabled_Selected_Image(
+                            cb_list_box_slider, Win_Find_Image("VSliderLargeThumbDisabled"));
```
Not "VSliderLargeThumbDisabledSelected" ?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                            cb_list_box_slider, Win_Find_Image("VSliderLargeHiliteSmallRepeatingCenter"));
+                        Gadget_Slider_Set_Thumb_Hilite_Image(cb_list_box_slider, Win_Find_Image("VSliderLargeThumbHilite"));
+                        Gadget_Slider_Set_Thumb_Hilite_Selected_Image(
+                            cb_list_box_slider, Win_Find_Image("VSliderLargeThumbHilite"));
```
Not "VSliderLargeThumbHiliteSelected" ?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                        Gadget_Button_Set_Enabled_Image(
+                            cb_list_box_slider_up_button, Win_Find_Image("VSliderLargeUpButtonEnabled"));
+                        Gadget_Button_Set_Enabled_Selected_Image(
+                            cb_list_box_slider_up_button, Win_Find_Image("VSliderLargeUpButtonEnabled"));
```
Not "VSliderLargeUpButtonEnabledSelected" ?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                        Gadget_Button_Set_Disabled_Image(
+                            cb_list_box_slider_up_button, Win_Find_Image("VSliderLargeUpButtonDisabled"));
+                        Gadget_Button_Set_Disabled_Selected_Image(
+                            cb_list_box_slider_up_button, Win_Find_Image("VSliderLargeUpButtonDisabled"));
```
Not "VSliderLargeUpButtonDisabledSelected" ?

jonwil replied:
This is correct.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+    if (window != nullptr) {
+        if ((window->m_status & WIN_STATUS_NO_INPUT) != 0) {
+            return nullptr;
+        } else if (ignore_enabled && (window->m_status & WIN_STATUS_ENABLED) == 0) {
```
This test looks redundant. Already tested 3 times above.

jonwil replied:
This is what the original does.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2594 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+
+void GameWindowManager::Register_Tab_List(std::list<GameWindow *> list)
+{
+    m_tabList.clear();
```
Call to clear obsolete.

xezon reviewed (commented):
First batch

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2597 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+
+    if (m_grabWindow != nullptr) {
+        switch (msg) {
+            case GWM_NONE:
```
Can move this atop `case GWM_LEFT_DRAG:` and remove goto label.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2597 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                    int delta_y = delta_pos->y;
+
+                    if (m_grabWindow->Win_Get_Parent() != nullptr) {
+                        GameWindow *window4 = m_grabWindow->Win_Get_Parent();
```
grab_window_parent

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2599 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+    WinInputReturnCode ret = WIN_INPUT_NOT_USED;
+    bool b = false;
+    GameWindow *window = nullptr;
+    GameWindow *window2 = nullptr;
```
tooltip_window

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2599 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+{
+    WinInputReturnCode ret = WIN_INPUT_NOT_USED;
+    bool b = false;
+    GameWindow *window = nullptr;
```
interacted_window

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2599 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                if ((window->m_status & (WIN_STATUS_BELOW | WIN_STATUS_ABOVE | WIN_STATUS_HIDDEN)) == 0
+                    && mouse_pos->x >= window->m_region.lo.x && mouse_pos->x <= window->m_region.hi.x
+                    && mouse_pos->y >= window->m_region.lo.y && mouse_pos->y <= window->m_region.hi.y) {
+                    GameWindow *window5 = window->Win_Point_In_Any_Child(mouse_pos->x, mouse_pos->y, true, true);
```
hovered_window

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2599 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+            if ((window->m_status & WIN_STATUS_ABOVE) != 0 && (window->m_status & WIN_STATUS_HIDDEN) == 0
+                && mouse_pos->x >= window->m_region.lo.x && mouse_pos->x <= window->m_region.hi.x
+                && mouse_pos->y >= window->m_region.lo.y && mouse_pos->y <= window->m_region.hi.y) {
+                GameWindow *window5 = window->Win_Point_In_Any_Child(mouse_pos->x, mouse_pos->y, true, true);
```
hovered_window

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2599 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                if ((window->m_status & WIN_STATUS_BELOW) != 0 && (window->m_status & WIN_STATUS_HIDDEN) == 0
+                    && mouse_pos->x >= window->m_region.lo.x && mouse_pos->x <= window->m_region.hi.x
+                    && mouse_pos->y >= window->m_region.lo.y && mouse_pos->y <= window->m_region.hi.y) {
+                    GameWindow *window5 = window->Win_Point_In_Any_Child(mouse_pos->x, mouse_pos->y, true, true);
```
hovered_window

jonwil replied:
fixed.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2599 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+                window2->m_instData.Get_Tooltip_Text(), window2->m_instData.m_tooltipDelay, 0, 1.0f);
+        }
+    } else {
+        b = true;
```
This bool is set here but never used. Mistake?

jonwil replied:
Seems to be unused from what I can see.

xezon commented on src/game/client/gui/gamewindowmanager.cpp:
```diff
@@ -106,3 +140,2599 @@ void GameWindowManager::Unlink_Child_Window(GameWindow *window)
+    GameWindow *window2 = nullptr;
+    bool clear_grab_window = false;
+    IRegion2D new_region;
+    ICoord2D grab_size;
```
Can move `new_region`, `grab_size` down to where used.

xezon commented:
Message Box buttons are incorrectly placed.

![shot_20231003_133744_2](https://github.com/TheAssemblyArmada/Thyme/assets/4720891/409759b5-b703-45f6-bd7b-afd712ac55c5)

Not clear to me how to narrow down. I suspect error in `GameWindowManager::Go_Go_Message_Box`

xezon commented:
Hovering image buttons in Control Bar will not correctly change the cursor and not show the button tool tip box.

I suspect error in `GameWindowManager::Win_Process_Mouse_Event`.

![shot_20231003_133729_1](https://github.com/TheAssemblyArmada/Thyme/assets/4720891/46269c86-df0f-487b-b1f1-815725ea7dae)

Trying to narrow down source of problem by commenting hooks will just make all image buttons disappear.

![shot_20231003_140431_1](https://github.com/TheAssemblyArmada/Thyme/assets/4720891/27f337be-ee4d-4ffb-95b6-0d2925618d72)

jonwil commented:
Fixed the issue with the missing tooltips, the issue with the incorrect cursor and the issue with message box buttons being in the wrong place. Also fixed an issue that could cause some controls to use the wrong draw function.

xezon commented:
Clicking on the Detail > Custom entry in the Options menu will not open the Detail Options.

![shot_20231004_174752_1](https://github.com/TheAssemblyArmada/Thyme/assets/4720891/971dd37e-6b72-44e6-ad6d-4d063fea827c)

jonwil commented:
Options menu fixed.
